### PR TITLE
Create a database for the users, create and implement the FireStore Repository

### PR DESCRIPTION
**Pull request by nada-ezziti**
**Additions** 
- Create a collection in FireStore to save the user's informations.
- Create the User Repository interface.
- Implement the Firestore Repository
- Create the User ViewModel
- Test the code with unit tests 

**Next steps**
- Use the implemented functions in the FriendListScreen to offer the possibility to send requests to friends and add them.
- Manage other user data such as their profile photo in the Settings Screen.

### CODE REVIEW

File: app/src/main/java/com/github/se/signify/model/user/SaveUserToFireStore.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    val usersCollection = db.collection(collectionPath)
+    val userDocRef = usersCollection.document(userId)
+
+    userDocRef
+        .get()
+        .addOnSuccessListener { document ->
+          if (!document.exists()) {
+            // The user does not exist yet, so we add them
+            userDocRef
+                .set(user, SetOptions.merge())
```
Comment (SaFouNaldo) [inline app/src/main/java/com/github/se/signify/model/user/SaveUserToFireStore.kt:41]: - Correctly handles both existing and non-existing users.
- Uses SetOptions.merge() to avoid overwriting existing fields.

Comment (SaFouNaldo) [reply app/src/main/java/com/github/se/signify/model/user/SaveUserToFireStore.kt:41]: For improved reliability, you might want to use documentSnapshot.exists() in combination with documentSnapshot.data to validate data consistency before setting it.

File: app/src/main/java/com/github/se/signify/model/user/User.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.github.se.signify.model.user
```
Comment (SaFouNaldo) [inline app/src/main/java/com/github/se/signify/model/user/User.kt:1]: - Clearly structured data model with appropriate default values, ensuring null-safety and easier Firestore serialization/deserialization.
- Use of emptyList() for friend-related lists avoids null issues.
- The User class seems well-defined and directly integrates into Firestore with minimal additional mapping.

File: app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package com.github.se.signify.model.user
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt:1]: Good coverage of the repository and ViewModel interactions.
Proper use of Mockito for mocking Firestore and Firebase dependencies.
Use of verify ensures that repository methods are correctly triggered.
RobolectricTestRunner and shadowOf help simulate Android’s Looper for asynchronous testing.
## Improvement Suggestions:
- Expand test coverage for edge cases, e.g., Firestore failures, or checking if users get added under various network conditions.
- Since Firestore operations are asynchronous, you could add timeouts to ensure Firebase responses are handled properly during tests.

**Review by SaFouNaldo — APPROVED** [review]
The code is well-organized and implements a clean architecture that separates concerns between Firebase operations, the repository, and the ViewModel. The use of StateFlow, Firebase's real-time listeners, and Firestore's array operations shows attention to performance and best practices. There are some areas that can be enhanced by reducing repetitive code and adding more robust testing around failure conditions, but overall, it's a solid implementation of Firebase-based user and friend management.

File: app/src/main/java/com/github/se/signify/model/user/UserRepositoryFireStore.kt [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    targetUserRef
+        .update(friendRequestsListPath, FieldValue.arrayUnion(currentUserId))
+        .addOnSuccessListener {
+          onSuccess() // Trigger onSuccess callback
+        }
+        .addOnFailureListener { e ->
+          onFailure(e) // Trigger onFailure callback
+        }
+  }
```
Comment (samymessouak) [inline app/src/main/java/com/github/se/signify/model/user/UserRepositoryFireStore.kt:116]: The UserRepositoryFireStore class provides various user management methods, including getFriendsList, sendFriendRequest, and acceptFriendRequest. The use of FieldValue.arrayUnion() and FieldValue.arrayRemove() for adding/removing friends or requests is a clean and efficient way to manage these list operations.

File: app/src/test/java/com/github/se/signify/model/user/SaveUserToFireStoreTest.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    // Act
+    saveUserToFireStore()
+
+    // Idle the main looper to process tasks
+    shadowOf(Looper.getMainLooper()).idle()
+
+    // Assert
+    `when`(mockDocumentReference.set(any(), eq(SetOptions.merge())))
+        .thenReturn(Tasks.forResult(null))
+  }
```
Comment (samymessouak) [inline app/src/test/java/com/github/se/signify/model/user/SaveUserToFireStoreTest.kt:94]: The unit tests in UserRepositoryFireStoreTest validate critical parts of the logic, such as adding a user if they don’t already exist, or verifying that existing users aren't added again. The use of mock objects and Robolectric ensures that interactions with Firebase can be tested without an actual backend. It’s important to confirm these tests cover all edge cases and potential errors.

**Review by samymessouak — APPROVED** [review]
PR Approved: Great work on this PR! 

The new functionalities look well-implemented:

- Creating the Firestore collection to manage user data is efficient and scalable.
- The repository interface and its Firestore implementation provide a solid foundation for managing user interactions.
- The new ViewModel integrates these functions smoothly, ensuring a clean separation of concerns.
- The use of FieldValue.arrayUnion() and FieldValue.arrayRemove() for list operations is effective, and I appreciate the addition of unit tests to ensure robustness. Next steps for using these features in the FriendListScreen are well-planned.

Overall, the changes are clear, and I didn't find any major issues. This should provide a strong base for upcoming features like sending friend requests and managing user data. Great job.
